flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/co/CoBroadcastWithKeyedOperator.java

```diff
@@ -0,0 +1,369 @@
+import org.apache.flink.streaming.api.operators.InternalTimer;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.operators.TimestampedCollector;
+import org.apache.flink.streaming.api.operators.Triggerable;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.OutputTag;
+import org.apache.flink.util.Preconditions;
+
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A {@link TwoInputStreamOperator} for executing {@link KeyedBroadcastProcessFunction}.
+ *
+ * @param <KS> The key type of the input keyed stream.
+ * @param <IN1> The input type of the broadcast side.
+ * @param <IN2> The input type of the keyed (non-broadcast) side.
+ * @param <K> The key type of the elements in the {@link BroadcastState}.
+ * @param <V> The value type of the elements in the {@link BroadcastState}.
+ * @param <OUT> The output type of the operator.
+ */
+@Internal
+public class CoBroadcastWithKeyedOperator<KS, IN1, IN2, K, V, OUT>
+		extends AbstractUdfStreamOperator<OUT, KeyedBroadcastProcessFunction<KS, IN1, IN2, K, V, OUT>>
+		implements TwoInputStreamOperator<IN1, IN2, OUT>, Triggerable<K, VoidNamespace> {
+
+	private static final long serialVersionUID = 5926499536290284870L;
+
+	private final MapStateDescriptor<K, V> broadcastStateDescriptor;
+
+	private transient TimestampedCollector<OUT> collector;
+
+	private transient BroadcastState<K, V> broadcastState;
+
+	private transient ReadWriteContextImpl rwContext;
+
+	private transient ReadOnlyContextImpl rContext;
+
+	private transient OnTimerContextImpl onTimerContext;
+
+	public CoBroadcastWithKeyedOperator(
+			final KeyedBroadcastProcessFunction<KS, IN1, IN2, K, V, OUT> function,
+			final MapStateDescriptor<K, V> broadcastStateDescriptor) {
+		super(function);
+		this.broadcastStateDescriptor = Preconditions.checkNotNull(broadcastStateDescriptor);
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		InternalTimerService<VoidNamespace> internalTimerService =
+				getInternalTimerService("user-timers", VoidNamespaceSerializer.INSTANCE, this);
+
+		TimerService timerService = new SimpleTimerService(internalTimerService);
+
+		collector = new TimestampedCollector<>(output);
+		broadcastState = getOperatorStateBackend().getBroadcastState(broadcastStateDescriptor);
+
+		rwContext = new ReadWriteContextImpl(getKeyedStateBackend(), userFunction, broadcastState, timerService);
+		rContext = new ReadOnlyContextImpl(userFunction, broadcastState, timerService);
+		onTimerContext = new OnTimerContextImpl(userFunction, broadcastState, timerService);
+	}
+
+	@Override
+	public void processElement1(StreamRecord<IN1> element) throws Exception {
```

Comment:
The operators still have the mapping `broadcast -> first input`, `main input -> second input`. This should be brought in line with how the API now works.